lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,181 @@
+        ): NadelSchemaValidationError? {
+            if (hydrationSourceFieldType.unwrapNonNull() is GraphQLScalarType && actorFieldArgType.unwrapNonNull() is GraphQLScalarType) {
+                if (isScalarAssignable(hydrationSourceFieldType.unwrapNonNull() as GraphQLScalarType, actorFieldArgType.unwrapNonNull() as GraphQLScalarType)) {
+                    return null
+                } else {
+                    return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                    )
+                }
+            }
+            return null
+        }
+
+        private fun isScalarAssignable(typeToAssign: GraphQLNamedInputType, targetType: GraphQLNamedInputType): Boolean {
+            if (typeToAssign.name == targetType.name) {
+                return true
+            }
+            if (targetType.name == Scalars.GraphQLID.name &&
+                    (typeToAssign.name == Scalars.GraphQLString.name || typeToAssign.name == Scalars.GraphQLInt.name)) {
+                return true
+            }
+            if (targetType is GraphQLInterfaceType && typeToAssign is GraphQLImplementingType) {
+                return typeToAssign.interfaces.contains(targetType)
+            }
+            return false
+        }
+
+        private fun validateListArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                hydration: UnderlyingServiceHydration
+        ): NadelSchemaValidationError? {
+            var hydrationSourceFieldInnerType: GraphQLType = hydrationSourceFieldType.unwrapNonNull().unwrapOne()
+            var actorFieldArgInnerType: GraphQLType = actorFieldArgType.unwrapNonNull().unwrapOne()
+            val errorExists = getHydrationInputErrors(
+                    hydrationSourceFieldInnerType,
+                    actorFieldArgInnerType,
+                    parent,
+                    overallField,
+                    remoteArg,
+                    hydration
+            ) != null
+            if (errorExists) {
+                return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydrationSourceFieldType,
+                        actorFieldArgType,
+                )
+            }
+            return null
+        }
+
+        private fun validateInputObjectArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
```

Comment:
```suggestion
                hydrationSourceFieldType: GraphQLObjectType,
                actorFieldArgType: GraphQLInputObjectType,
```

Per above comment about passing in smart casted objects